packages/front-end/components/PowerCalculation/stats.ts

```diff
@@ -340,3 +356,294 @@ export function powerMetricWeeks(
+  return numerator / denominator;
+}
+
+/**
+ * Performs power calculation
+ *
+ * @param metric an object of class MetricParams that has info about prior mean and sd, metric mean and sd, and effect size.
+ * @param alpha false positive rate (default: 0.05).
+ * @param nPerVariation sample size per variation.
+ * @param relative boolean indicator if relative inference is desired.
+ * @returns Estimated power.
+ */
+export function powerEstBayesian(
+  metric: MetricParams,
+  alpha: number,
+  nPerVariation: number,
+  relative: boolean
+): number {
+  const upperCutpoint = getCutpoint(
+    metric,
+    alpha,
+    nPerVariation,
+    relative,
+    true
+  );
+  const lowerCutpoint = getCutpoint(
+    metric,
+    alpha,
+    nPerVariation,
+    relative,
+    false
+  );
+  const powerPos = 1.0 - normal.cdf(upperCutpoint, 0, 1);
+  const powerNeg = normal.cdf(lowerCutpoint, 0, 1);
+  return powerPos + powerNeg;
+}
+
+/**
+ * Performs mde calc
+ * @param metric an object of class MetricParams that has info about prior mean and sd, metric mean and sd, and effect size.
+ * @param alpha false positive rate (default: 0.05).
+ * @param nPerVariation sample size per variation.
+ * @param relative boolean indicator if relative inference is desired.
+ * @param stepSize step size used in initial grid search.
+ * @returns object of class MDEResults, containing type and either mde or description.
+ */
+function sweepGridFine(
+  metric: MetricParams,
+  alpha: number,
+  power: number,
+  nPerVariation: number,
+  relative: boolean,
+  stepSize: number
+): MDEResults {
+  const effectSize = metric.effectSize;
+  const stepSizeFiner = stepSize / 100;
+  const lowerBound = effectSize - stepSize;
+  for (
+    let effectSizeFiner = lowerBound;
+    effectSizeFiner < effectSize;
+    effectSizeFiner += stepSizeFiner
+  ) {
+    metric.effectSize = effectSizeFiner;
+    const p = powerEstBayesian(metric, alpha, nPerVariation, relative);
+    if (p >= power) {
+      const mdeResults: MDEResults = {
+        type: "success",
+        mde: effectSizeFiner,
+      };
+      return mdeResults;
+    }
+  }
+  const mdeResults: MDEResults = {
+    type: "error",
+    description: "MDE achieving power = 0.8 does not exist in this range. ",
+  };
+  return mdeResults;
+}
+
+/**
+ * Performs mde calc
+ * @param metric an object of class MetricParams that has info about prior mean and sd, metric mean and sd, and effect size.
+ * @param alpha false positive rate (default: 0.05).
+ * @param nPerVariation sample size per variation.
+ * @param relative boolean indicator if relative inference is desired.
+ * @returns object of class MDEResults, containing type and either mde or description.
+ */
+export function findMdeBayesian(
+  metric: MetricParams,
+  alpha: number,
+  power: number,
+  nPerVariation: number,
+  relative: boolean
+): MDEResults {
+  /*fixed effect size, so prior variance of data generating process is 0*/
+  const dummyMetric = { ...metric };
```

Comment:
I think a safer way to do this would to do :

`const dummyMetric = { ...metric, effectSize: effectSize}` so that you're not overwriting the shallow copy but just building it with the new value.

Reply:
done, thanks for the tip